File: src/sync.ts
import { spawnSync, spawn } from "child_process";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

export function syncClaudeConfig(containerName: string): void {
  const claudeDir = path.join(os.homedir(), ".claude");
  const containerClaudeDir = "/home/node/.claude";
  const settingsFile = path.join(claudeDir, "settings.json");
  const claudeMdFile = path.join(claudeDir, "CLAUDE.md");

  const filesToCopy: string[] = [];
  let containerSettings: Record<string, unknown> = {};

  if (fs.existsSync(settingsFile)) {
    try {
      const content = fs.readFileSync(settingsFile, "utf-8");
      const settings = JSON.parse(content);

      if (settings.statusLine) {
        containerSettings.statusLine = settings.statusLine;

        if (settings.statusLine.command) {
          let cmdPath = settings.statusLine.command;
          if (cmdPath.startsWith("~")) {
            cmdPath = os.homedir() + cmdPath.slice(1);
          }
          if (fs.existsSync(cmdPath)) {
            filesToCopy.push(cmdPath);
          }
        }
      }
    } catch (e) {
      console.log(`Warning: Failed to parse settings.json: ${e}`);
    }
  }

  spawnSync(
    "docker",
    ["exec", containerName, "mkdir", "-p", containerClaudeDir],
    { stdio: ["pipe", "pipe", "pipe"] }
  );

  if (fs.existsSync(claudeMdFile)) {
    filesToCopy.push(claudeMdFile);
  } else {
    spawnSync(
      "docker",
      ["exec", containerName, "rm", "-f", `${containerClaudeDir}/CLAUDE.md`],
      { stdio: ["pipe", "pipe", "pipe"] }
    );
  }

  const commandsDir = path.join(claudeDir, "commands");
  let commandsSource: string | null = null;
  if (fs.existsSync(commandsDir)) {
    try {
      commandsSource = fs.realpathSync(commandsDir);
      if (!fs.statSync(commandsSource).isDirectory()) {
        commandsSource = null;
      }
    } catch {
      commandsSource = null;
    }
  }

  if (commandsSource) {
    spawnSync(
      "docker",
      ["exec", containerName, "rm", "-rf", `${containerClaudeDir}/commands`],
      { stdio: ["pipe", "pipe", "pipe"] }
    );
    spawnSync(
      "docker",
      ["exec", containerName, "mkdir", "-p", `${containerClaudeDir}/commands`],
      { stdio: ["pipe", "pipe", "pipe"] }
    );

    const tarCreate = spawn("tar", ["-cf", "-", "."], {
      cwd: commandsSource,
      stdio: ["pipe", "pipe", "pipe"],
    });
    const tarExtract = spawn(
      "docker",
      [
        "exec",
        "-i",
        containerName,
        "tar",
        "-xf",
        "-",
        "-C",
        `${containerClaudeDir}/commands`,
      ],
      { stdio: ["pipe", "pipe", "pipe"] }
    );
    tarCreate.stdout.pipe(tarExtract.stdin);

    tarCreate.on("close", () => {});
    tarExtract.on("close", (code) => {
      if (code !== 0) {
        console.log("Warning: Failed to copy commands directory");
      }
    });
  } else {
    spawnSync(
      "docker",
      ["exec", containerName, "rm", "-rf", `${containerClaudeDir}/commands`],
      { stdio: ["pipe", "pipe", "pipe"] }
    );
  }

  if (filesToCopy.length > 0) {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "vibecon-"));
    try {
      for (const srcFile of filesToCopy) {
        const destPath = path.join(tmpDir, path.basename(srcFile));
        fs.copyFileSync(srcFile, destPath);
        try {
          const srcStats = fs.statSync(srcFile);
          if (srcStats.mode & 0o111) {
            fs.chmodSync(destPath, srcStats.mode | 0o111);
          }
        } catch {
          // Ignore permission errors
        }
      }

      const tarCreate = spawn("tar", ["-cf", "-", "."], {
        cwd: tmpDir,
        stdio: ["pipe", "pipe", "pipe"],
      });
      const tarExtract = spawn(
        "docker",
        [
          "exec",
          "-i",
          containerName,
          "tar",
          "-xf",
          "-",
          "-C",
          containerClaudeDir,
        ],
        { stdio: ["pipe", "pipe", "pipe"] }
      );
      tarCreate.stdout.pipe(tarExtract.stdin);
    } finally {
      setTimeout(() => {
        try {
          fs.rmSync(tmpDir, { recursive: true });
        } catch {
          // Ignore cleanup errors
        }
      }, 1000);
    }
  }

  if (Object.keys(containerSettings).length > 0) {
    const settingsJson = JSON.stringify(containerSettings, null, 2);
    spawnSync(
      "docker",
      [
        "exec",
        containerName,
        "sh",
        "-c",
        `cat > ${containerClaudeDir}/settings.json << 'EOFCONFIG'\n${settingsJson}\nEOFCONFIG`,
      ],
      { stdio: ["pipe", "pipe", "pipe"] }
    );
  }

  spawnSync(
    "docker",
    [
      "exec",
      "-u",
      "root",
      containerName,
      "chown",
      "-R",
      "node:node",
      containerClaudeDir,
    ],
    { stdio: ["pipe", "pipe", "pipe"] }
  );
}

File: internal/sync/sync.go
package sync

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const (
	containerClaudeDir = "/home/node/.claude"
)

// ClaudeSettings represents the subset of Claude settings we sync
type ClaudeSettings struct {
	StatusLine map[string]interface{} `json:"statusLine,omitempty"`
}

// SyncClaudeConfig syncs Claude config to container
func SyncClaudeConfig(containerName string) error {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("failed to get home directory: %w", err)
	}

	claudeDir := filepath.Join(homeDir, ".claude")
	settingsFile := filepath.Join(claudeDir, "settings.json")
	claudeMdFile := filepath.Join(claudeDir, "CLAUDE.md")

	// Track files to copy
	var filesToCopy []string
	containerSettings := ClaudeSettings{}

	// Parse settings.json if it exists
	if _, err := os.Stat(settingsFile); err == nil {
		data, err := os.ReadFile(settingsFile)
		if err == nil {
			var settings map[string]interface{}
			if err := json.Unmarshal(data, &settings); err == nil {
				// Extract statusLine section if present
				if statusLine, ok := settings["statusLine"].(map[string]interface{}); ok {
					containerSettings.StatusLine = statusLine

					// If statusLine has a command, add that file to copy list
					if cmdPath, ok := statusLine["command"].(string); ok {
						// Expand ~ to home directory
						if strings.HasPrefix(cmdPath, "~") {
							cmdPath = filepath.Join(homeDir, cmdPath[1:])
						}
						if _, err := os.Stat(cmdPath); err == nil {
							filesToCopy = append(filesToCopy, cmdPath)
						}
					}
				}
			}
		}
	}

	// Ensure container directory exists
	cmd := exec.Command("docker", "exec", containerName, "mkdir", "-p", containerClaudeDir)
	cmd.Stdout = nil
	cmd.Stderr = nil
	_ = cmd.Run()

	// Handle CLAUDE.md sync
	if _, err := os.Stat(claudeMdFile); err == nil {
		filesToCopy = append(filesToCopy, claudeMdFile)
	} else {
		// Remove from container if doesn't exist locally
		cmd := exec.Command("docker", "exec", containerName, "rm", "-f",
			filepath.Join(containerClaudeDir, "CLAUDE.md"))
		cmd.Stdout = nil
		cmd.Stderr = nil
		_ = cmd.Run()
	}

	// Handle commands directory sync
	commandsDir := filepath.Join(claudeDir, "commands")
	commandsSource := ""
	if info, err := os.Lstat(commandsDir); err == nil {
		if info.Mode()&os.ModeSymlink != 0 {
			// Resolve symlink
			target, err := filepath.EvalSymlinks(commandsDir)
			if err == nil {
				if stat, err := os.Stat(target); err == nil && stat.IsDir() {
					commandsSource = target
				}
			}
		} else if info.IsDir() {
			commandsSource = commandsDir
		}
	}

	if commandsSource != "" {
		// Remove existing commands directory
		cmd := exec.Command("docker", "exec", containerName, "rm", "-rf",
			filepath.Join(containerClaudeDir, "commands"))
		cmd.Stdout = nil
		cmd.Stderr = nil
		_ = cmd.Run()

		// Create fresh commands directory
		cmd = exec.Command("docker", "exec", containerName, "mkdir", "-p",
			filepath.Join(containerClaudeDir, "commands"))
		cmd.Stdout = nil
		cmd.Stderr = nil
		_ = cmd.Run()

		// Copy commands directory using tar
		if err := copyDirToContainer(commandsSource, containerName,
			filepath.Join(containerClaudeDir, "commands")); err != nil {
			fmt.Fprintf(os.Stderr, "Warning: Failed to copy commands directory: %v\n", err)
		}
	} else {
		// Remove commands directory from container
		cmd := exec.Command("docker", "exec", containerName, "rm", "-rf",
			filepath.Join(containerClaudeDir, "commands"))
		cmd.Stdout = nil
		cmd.Stderr = nil
		_ = cmd.Run()
	}

	// Copy files using tar if we have any
	if len(filesToCopy) > 0 {
		if err := copyFilesToContainer(filesToCopy, containerName, containerClaudeDir); err != nil {
			fmt.Fprintf(os.Stderr, "Warning: Failed to copy files: %v\n", err)
		}
	}

	// Write container settings.json if we have settings to write
	if containerSettings.StatusLine != nil {
		settingsJSON, err := json.MarshalIndent(containerSettings, "", "  ")
		if err == nil {
			// Write settings using shell heredoc
			shellCmd := fmt.Sprintf("cat > %s/settings.json << 'EOFCONFIG'\n%s\nEOFCONFIG",
				containerClaudeDir, string(settingsJSON))
			cmd := exec.Command("docker", "exec", containerName, "sh", "-c", shellCmd)
			cmd.Stdout = nil
			_ = cmd.Run()
		}
	}

	// Fix ownership for node user
	cmd = exec.Command("docker", "exec", "-u", "root", containerName,
		"chown", "-R", "node:node", containerClaudeDir)
	cmd.Stdout = nil
	cmd.Stderr = nil
	_ = cmd.Run()

	return nil
}

// copyFilesToContainer copies files to container using tar
func copyFilesToContainer(files []string, containerName, targetDir string) error {
	// Create a temporary directory with just the filenames
	tmpDir, err := os.MkdirTemp("", "vibecon-sync-*")
	if err != nil {
		return fmt.Errorf("failed to create temp dir: %w", err)
	}
	defer os.RemoveAll(tmpDir)

	// Copy files to temp dir with just their basenames
	for _, srcFile := range files {
		data, err := os.ReadFile(srcFile)
		if err != nil {
			continue
		}
		destFile := filepath.Join(tmpDir, filepath.Base(srcFile))
		if err := os.WriteFile(destFile, data, 0644); err != nil {
			continue
		}
		// Preserve executable bit
		if info, err := os.Stat(srcFile); err == nil {
			if info.Mode()&0111 != 0 {
				os.Chmod(destFile, 0755)
			}
		}
	}

	return copyDirToContainer(tmpDir, containerName, targetDir)
}

// copyDirToContainer copies a directory to container using tar
func copyDirToContainer(srcDir, containerName, targetDir string) error {
	// Create tar archive
	tarCmd := exec.Command("tar", "-cf", "-", ".")
	tarCmd.Dir = srcDir
	tarOut, err := tarCmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("failed to create tar pipe: %w", err)
	}

	// Extract in container
	extractCmd := exec.Command("docker", "exec", "-i", containerName,
		"tar", "-xf", "-", "-C", targetDir)
	extractCmd.Stdin = tarOut

	var extractErr bytes.Buffer
	extractCmd.Stderr = &extractErr

	if err := tarCmd.Start(); err != nil {
		return fmt.Errorf("failed to start tar: %w", err)
	}

	if err := extractCmd.Run(); err != nil {
		return fmt.Errorf("failed to extract tar: %s", extractErr.String())
	}

	if err := tarCmd.Wait(); err != nil {
		return fmt.Errorf("tar command failed: %w", err)
	}

	return nil
}
